File: src/chmod.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <getopt.h>

/*
 * chmod — change file mode bits
 *
 * Usage: chmod [-R] [-v] [-c] MODE FILE...
 *   -R  apply recursively
 *   -v  verbose: show each file processed
 *   -c  report only when a change is made
 *
 * MODE may be an octal number (e.g. 755) or a symbolic expression
 * (e.g. u+x, a=rw, go-w).
 */

static int flag_recursive = 0;
static int flag_verbose   = 0;
static int flag_changes   = 0;

/* ---- symbolic mode parser ------------------------------------------------ */

/* Return the 3-slot mask for a who character ('u','g','o','a') or -1. */
static int who_mask(char c)
{
    switch (c) {
    case 'u': return 0700;
    case 'g': return 0070;
    case 'o': return 0007;
    case 'a': return 0777;
    default:  return -1;
    }
}

/* Return the spread-across-all-slots bit pattern for a permission char,
 * or -1 if not a simple permission character (r/w/x/s/t). */
static int perm_bits(char c)
{
    switch (c) {
    case 'r': return 0444;
    case 'w': return 0222;
    case 'x': return 0111;
    case 's': return (int)(S_ISUID | S_ISGID);
    case 't': return (int)S_ISVTX;
    default:  return -1;
    }
}

/* Apply one symbolic clause (e.g. "u+x", "go-w", "a=rw") to *mode.
 * Returns 0 on success, -1 on parse error. */
static int apply_clause(mode_t *mode, const char *clause)
{
    const char *p = clause;
    int who = 0;

    while (*p && who_mask(*p) >= 0) {
        who |= who_mask(*p);
        p++;
    }
    if (who == 0)
        who = 0777; /* default: all */

    if (!*p || (*p != '+' && *p != '-' && *p != '='))
        return -1;

    char op = *p++;
    int perms = 0;

    while (*p) {
        int b = perm_bits(*p);
        if (b >= 0) {
            perms |= b;
        } else {
            /* Reference: copy bits from u/g/o source */
            int src = who_mask(*p);
            if (src < 0)
                return -1;
            mode_t src_bits = *mode & (mode_t)src;
            int shift = 0;
            if (src == 0700)      shift = 6;
            else if (src == 0070) shift = 3;
            perms |= (int)((src_bits >> shift) & 7) * 0111;
        }
        p++;
    }

    /* X — execute/search only if directory or already executable */
    /* (handled as a special clause prefix in parse_symbolic) */

    int masked = perms & who;
    switch (op) {
    case '+': *mode |=  (mode_t)masked; break;
    case '-': *mode &= ~(mode_t)masked; break;
    case '=': *mode  = (*mode & ~(mode_t)who) | (mode_t)masked; break;
    }
    return 0;
}

/* Apply a comma-separated symbolic mode string.  Returns 0 or -1. */
static int parse_symbolic(const char *spec, mode_t *mode)
{
    char buf[256];
    strncpy(buf, spec, sizeof(buf) - 1);
    buf[sizeof(buf) - 1] = '\0';

    char *save, *tok;
    for (tok = strtok_r(buf, ",", &save); tok;
         tok = strtok_r(NULL, ",", &save)) {
        if (apply_clause(mode, tok) != 0)
            return -1;
    }
    return 0;
}

/* Attempt to parse an octal mode; returns (mode_t)-1 on failure. */
static mode_t try_octal(const char *s)
{
    if (!*s) return (mode_t)-1;
    char *end;
    long val = strtol(s, &end, 8);
    if (*end != '\0' || val < 0 || val > 07777)
        return (mode_t)-1;
    return (mode_t)val;
}

/* ---- core operation ------------------------------------------------------ */

static int do_chmod(const char *path, const char *mode_str)
{
    mode_t octal = try_octal(mode_str);

    /* Fast path: pure octal mode, no feedback needed.
     * A single chmod() call avoids any TOCTOU window. */
    if (octal != (mode_t)-1 && !flag_verbose && !flag_changes) {
        if (chmod(path, octal) != 0) {
            fprintf(stderr, "chmod: changing permissions of '%s': %s\n",
                    path, strerror(errno));
            return 1;
        }
        return 0;
    }

    /* For symbolic modes or verbose output, open the file and operate on the
     * file descriptor so that fstat() and fchmod() reference the same inode,
     * eliminating the TOCTOU window between the stat and the chmod. */
    int fd = open(path, O_RDONLY | O_NONBLOCK);
    if (fd < 0) {
        fprintf(stderr, "chmod: cannot open '%s': %s\n", path, strerror(errno));
        return 1;
    }

    struct stat st;
    if (fstat(fd, &st) != 0) {
        fprintf(stderr, "chmod: cannot stat '%s': %s\n", path, strerror(errno));
        close(fd);
        return 1;
    }

    mode_t old_mode = st.st_mode & 07777;
    mode_t new_mode = (octal != (mode_t)-1) ? octal : old_mode;

    if (octal == (mode_t)-1) {
        /* Symbolic mode: compute new_mode from old_mode */
        if (parse_symbolic(mode_str, &new_mode) != 0) {
            fprintf(stderr, "chmod: invalid mode: '%s'\n", mode_str);
            close(fd);
            return 1;
        }
    }

    if (fchmod(fd, new_mode) != 0) {
        fprintf(stderr, "chmod: changing permissions of '%s': %s\n",
                path, strerror(errno));
        close(fd);
        return 1;
    }
    close(fd);

    if (flag_verbose || (flag_changes && old_mode != new_mode)) {
        if (old_mode != new_mode)
            printf("mode of '%s' changed from %04o to %04o\n",
                   path, old_mode, new_mode);
        else if (flag_verbose)
            printf("mode of '%s' retained as %04o\n", path, old_mode);
    }
    return 0;
}

static int chmod_tree(const char *path, const char *mode_str);

static int descend_dir(const char *path, const char *mode_str)
{
    DIR *d = opendir(path);
    if (!d) {
        fprintf(stderr, "chmod: cannot open directory '%s': %s\n",
                path, strerror(errno));
        return 1;
    }
    int rc = 0;
    struct dirent *de;
    while ((de = readdir(d)) != NULL) {
        if (strcmp(de->d_name, ".") == 0 || strcmp(de->d_name, "..") == 0)
            continue;
        char child[4096];
        snprintf(child, sizeof(child), "%s/%s", path, de->d_name);
        rc |= chmod_tree(child, mode_str);
    }
    closedir(d);
    return rc;
}

static int chmod_tree(const char *path, const char *mode_str)
{
    struct stat st;
    if (lstat(path, &st) != 0) {
        fprintf(stderr, "chmod: cannot stat '%s': %s\n", path, strerror(errno));
        return 1;
    }
    int rc = do_chmod(path, mode_str);
    if (S_ISDIR(st.st_mode))
        rc |= descend_dir(path, mode_str);
    return rc;
}

/* ---- main ---------------------------------------------------------------- */

static void usage(const char *prog)
{
    fprintf(stderr, "Usage: %s [-R] [-v] [-c] MODE FILE...\n", prog);
    exit(1);
}

int main(int argc, char *argv[])
{
    int opt;
    while ((opt = getopt(argc, argv, "Rvc")) != -1) {
        switch (opt) {
        case 'R': flag_recursive = 1; break;
        case 'v': flag_verbose   = 1; break;
        case 'c': flag_changes   = 1; break;
        default:  usage(argv[0]);
        }
    }

    if (optind + 1 >= argc)
        usage(argv[0]);

    const char *mode_str = argv[optind++];

    int rc = 0;
    for (; optind < argc; optind++) {
        if (flag_recursive)
            rc |= chmod_tree(argv[optind], mode_str);
        else
            rc |= do_chmod(argv[optind], mode_str);
    }
    return rc;
}

File: login_tools/cli/chmod.py
"""
chmod: Change file mode bits.

Usage: chmod [-R] [-v] [-c] MODE FILE...
"""
from __future__ import annotations

import argparse
import stat
import sys
from pathlib import Path


def _apply_clause(current: int, clause: str) -> int:
    """Apply one symbolic mode clause (e.g. u+x, go-w, a=r) to current mode bits."""
    i = 0
    who_chars = ''
    while i < len(clause) and clause[i] in 'ugoa':
        who_chars += clause[i]
        i += 1

    if i >= len(clause) or clause[i] not in '+-=':
        raise ValueError(f'invalid mode clause: {clause!r}')

    op = clause[i]
    perm_chars = clause[i + 1:]

    # Determine the 'who' mask (which permission slots to affect)
    who_bits = 0
    if not who_chars or 'a' in who_chars:
        who_bits = 0o777
    if 'u' in who_chars:
        who_bits |= 0o700
    if 'g' in who_chars:
        who_bits |= 0o070
    if 'o' in who_chars:
        who_bits |= 0o007

    # Determine permission bits to apply (always spread across all three slots)
    perm_bits = 0
    for c in perm_chars:
        if c == 'r':
            perm_bits |= 0o444
        elif c == 'w':
            perm_bits |= 0o222
        elif c == 'x':
            perm_bits |= 0o111
        elif c == 'X':
            # Set execute only if the path is a directory or already has any execute bit
            if stat.S_ISDIR(current) or (current & 0o111):
                perm_bits |= 0o111
        elif c == 's':
            perm_bits |= (stat.S_ISUID | stat.S_ISGID)
        elif c == 't':
            perm_bits |= stat.S_ISVTX
        elif c == 'u':
            user_bits = (current & 0o700) >> 6
            perm_bits |= user_bits * 0o111
        elif c == 'g':
            grp_bits = (current & 0o070) >> 3
            perm_bits |= grp_bits * 0o111
        elif c == 'o':
            oth_bits = current & 0o007
            perm_bits |= oth_bits * 0o111
        else:
            raise ValueError(f'invalid permission character: {c!r}')

    # Special bits (SUID, SGID, SVTX) are outside the 0o777 'who' mask and
    # must be handled separately so they are not accidentally cleared.
    # Mapping: u → SUID, g → SGID, o/a → SVTX (traditional convention)
    if not who_chars or 'a' in who_chars:
        special_who = stat.S_ISUID | stat.S_ISGID | stat.S_ISVTX
    else:
        special_who = 0
        if 'u' in who_chars:
            special_who |= stat.S_ISUID
        if 'g' in who_chars:
            special_who |= stat.S_ISGID
        if 'o' in who_chars:
            special_who |= stat.S_ISVTX
    special_masked = perm_bits & special_who

    masked = perm_bits & who_bits

    if op == '+':
        return current | masked | special_masked
    elif op == '-':
        return current & ~(masked | special_masked)
    else:  # '='
        return (current & ~(who_bits | special_who)) | masked | special_masked


def _apply_symbolic_mode(current: int, spec: str) -> int:
    """Apply a comma-separated symbolic mode string to current mode bits."""
    for clause in spec.split(','):
        current = _apply_clause(current, clause)
    return current


def _parse_mode(mode_str: str, current: int) -> int:
    """
    Parse a mode string (octal or symbolic) and return the resulting mode.
    For octal modes the current mode is ignored; for symbolic it is needed.
    """
    try:
        return int(mode_str, 8)
    except ValueError:
        pass
    return _apply_symbolic_mode(current, mode_str)


def _chmod_path(path: Path, mode_str: str, verbose: bool, changes: bool) -> int:
    """Apply chmod to a single path. Returns 0 on success, 1 on error."""
    try:
        st = path.stat()
        old_mode = st.st_mode & 0o7777
        new_mode = _parse_mode(mode_str, st.st_mode)
        path.chmod(new_mode)
        if verbose or (changes and old_mode != new_mode):
            if old_mode != new_mode:
                print(f"mode of '{path}' changed from {old_mode:04o} to {new_mode:04o}")
            elif verbose:
                print(f"mode of '{path}' retained as {old_mode:04o}")
        return 0
    except (OSError, ValueError) as exc:
        print(f'chmod: {path}: {exc}', file=sys.stderr)
        return 1


def _chmod_recursive(path: Path, mode_str: str, verbose: bool, changes: bool) -> int:
    rc = _chmod_path(path, mode_str, verbose, changes)
    if not path.is_symlink() and path.is_dir():
        for child in path.iterdir():
            rc |= _chmod_recursive(child, mode_str, verbose, changes)
    return rc


def main() -> None:
    parser = argparse.ArgumentParser(
        prog='chmod',
        description='Change file mode bits.',
    )
    parser.add_argument(
        'mode',
        help='File mode: octal (e.g. 755) or symbolic (e.g. u+x, a=rw)',
    )
    parser.add_argument('files', nargs='+', metavar='FILE', help='Files to modify')
    parser.add_argument(
        '-R', '--recursive',
        action='store_true',
        help='Change files and directories recursively',
    )
    parser.add_argument(
        '-v', '--verbose',
        action='store_true',
        help='Output a diagnostic for every file processed',
    )
    parser.add_argument(
        '-c', '--changes',
        action='store_true',
        help='Like --verbose but report only when a change is made',
    )
    args = parser.parse_args()

    rc = 0
    for f in args.files:
        p = Path(f)
        if not p.exists() and not p.is_symlink():
            print(
                f"chmod: cannot access '{f}': No such file or directory",
                file=sys.stderr,
            )
            rc = 1
            continue
        if args.recursive:
            rc |= _chmod_recursive(p, args.mode, args.verbose, args.changes)
        else:
            rc |= _chmod_path(p, args.mode, args.verbose, args.changes)
    sys.exit(rc)


if __name__ == '__main__':
    main()
